CMake build system support

hoijui opened the pull request:
adds basic support for building and installing CMatrix through CMake (a cross-platform meta-build system).
the binary should compiled on all platforms (untested), but the fonts are only installed on Unix systems.

hoijui committed:
add support to build with CMake (a meta build system)

use like this:

	mkdir -p build
	cd build
	cmake ..
	# or
    	#cmake -DCMAKE_INSTALL_PREFIX=/usr ..
	make

abishekvashok commented on cmatrix.c:
```diff
@@ -32,8 +32,6 @@
 #include <termios.h>
 #include <signal.h>
 
-#include "config.h"
```
config.h is generated using make and is needed for some configuration options that we will leave for the user.

abishekvashok reviewed (changes requested):
Please check the failing tests in Travis as well.

hoijui commented:
i added a simple fix for config.h

abishekvashok commented:
@hoijui can you modify the .travis.yml file as well so that we can make sure everything works well there as well.

hoijui commented:
ok.. did that, and it seems to work :-)
is it ok that now configure is not tested anymore? 
i never used travis before.. did not find out how to do the two of them (i guess that would mean, having two scripts)

abishekvashok commented:
You can parallise the build. It's important to run tests on configure.

https://docs.travis-ci.com/user/speeding-up-the-build/

If you can't find help. I would help on this.

hoijui commented:
thanks for the hint!
looks like now i got it right

abishekvashok commented on .travis.yml:
```diff
@@ -4,9 +4,18 @@ os: linux
 script:
-  - ./configure
+  - mkdir -p build
+  - cd build
```
Cd outside the build and then run configure

hoijui replied:
so with configure, we can not make an out-of-source build?
why not.. and can we "fix" that?

abishekvashok replied:
Basically the configuration file resides in the root directory of the project as a standard. Since you cd into build, it won't be there in the build folder and you are running ./configure which will make the shell to execute the configure file in ./ i.e, current folder(build).

hoijui commented:
ahh...
when you look at the file as the last commit of this bunch leaves it, it looks like this:
https://github.com/hoijui/cmatrix/blob/8c8ae0b4c3b8ebf102e0ec2275e1a0b635591757/.travis.yml
it calls `../configure`, and the tests ran fine, and did actually execute both the configure and the cmake way of doing things.

abishekvashok commented:
Ah my bad, i will test this my local setup and let you know.

abishekvashok commented:
Yes it works. Thanks for that!

abishekvashok commented:
@hoijui could you provide a nice commit message so that I could squash and merge or just squash those commits

hoijui commented:
... ok! i "beautified" the commit history now, leaving only two commits, which i though make sense separately. i checked that the content is still the same with `git diff`.

(old history is now under https://github.com/hoijui/cmatrix/tree/cmake-develop)

hoijui commented:
ouhh.. now thinking about it, should i also add something to README.md maybe?

hoijui committed:
README.md: use a more accurate sub-heading [minor]

hoijui committed:
README.md: fix build instructions [fix]

hoijui committed:
README.md: use more standard conform syntax [minor]

This allows for easy copy&paste of build instructions,
and also generates nice output with pandoc now.
You may test that with:

    pandoc -f markdown -t HTML README.md > README.html

hoijui committed:
README.md: explain how to build and install with cmake [minor]

hoijui commented:
i did that, adding 4 READE modifying commits, of which the first 3 are independent of cmake, but for now i included them in this branch anyway. please say if i should separate them out into an other branch/pull request. they are minor though.
also say if i should squash the cmake related README commit into the main cmake commit.

abishekvashok commented:
Please separate them out. They should go on another issue with another pr. Please, revert those commits, they are causing build errors as well.

hoijui commented:
actually, they are not producing the build errors, but the instability of the _config_ build method is. i noted this a few times already, with this very same error (`aclocal` version mismatch). it seems to pop up randomly, and the way to "fix" it, is to just clean and rebuild.

hoijui commented:
it would be easiest if the README related pull request would be merge in first, and then these cmake changes afterwards.

abishekvashok commented:
@hoijui done.

abishekvashok committed:
Merge branch 'master' into cmake

abishekvashok committed:
Merge branch 'master' into cmake

hoijui commented:
thanks for the management work. :-)
from my point of view, this branch could be merged now.

abishekvashok commented:
@hoijui would be happy if you could provide a nice descriptiuon as the commits will be squashed.

hoijui commented:
you mean, you want a commit message for the merge commit, or you do not want to merge, but have a single, rebased, squashed commit?
if the second: why?
because i think, code history should be preserved and be as detailed and divided as possible. when it comes in as a merge of an appropriately named branch, i see no disadvantage for that approach, but many benefits.

a description can be taken from this commit:
https://github.com/abishekvashok/cmatrix/pull/17/commits/9fffb25f1414d7b73c423714feb512de701624ef

abishekvashok commented:
I think you are right.

hoijui commented:
:o wow. cool.. thanks!
really great doing "business" with you! :-)
